src/CoreWCF.Primitives/src/CoreWCF/Channels/IReliableChannelBinder.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System.Threading.Tasks;
+using System;
+
+namespace CoreWCF.Channels
+{
+    internal delegate void BinderExceptionHandler(IReliableChannelBinder sender, Exception exception);
+
+    internal interface IReliableChannelBinder
+    {
+        IChannel Channel { get; }
+        bool Connected { get; }
+        TimeSpan DefaultSendTimeout { get; }
+        bool HasSession { get; }
+        EndpointAddress LocalAddress { get; }
+        EndpointAddress RemoteAddress { get; }
+        CommunicationState State { get; }
+
+        event BinderExceptionHandler Faulted;
+        event BinderExceptionHandler OnException;
+
+        void Abort();
+
+        Task CloseAsync(TimeSpan timeout);
+        Task CloseAsync(TimeSpan timeout, MaskingMode maskingMode);
+        Task OpenAsync(TimeSpan timeout);
+        Task SendAsync(Message message, TimeSpan timeout);
+        Task SendAsync(Message message, TimeSpan timeout, MaskingMode maskingMode);
+
+        Task<(bool, RequestContext)> TryReceiveAsync(TimeSpan timeout);
+        Task<(bool, RequestContext)> TryReceiveAsync(TimeSpan timeout, MaskingMode maskingMode);
```

Comment:
TryReceiveAsync is still a pull model. I haven't got to the implementation part yet, but initially this seems incorrect. I would expect maybe a DispatchAsync method which gets passed a RequestContext. I'll try to come back to this later, but giving a heads up now.